pkg/lokifrontend/frontend/v2/frontend_test.go

```diff
@@ -252,6 +251,94 @@ func TestFrontendFailedCancellation(t *testing.T) {
 	})
 }
 
+func TestFrontendStoppingWaitsForEmptyInflightRequests(t *testing.T) {
+	delayResponse := 10 * time.Millisecond
+	f, _ := setupFrontend(t, func(f *Frontend, msg *schedulerpb.FrontendToScheduler) *schedulerpb.SchedulerToFrontend {
+		// We cannot call QueryResult directly, as Frontend is not yet waiting for the response.
+		// It first needs to be told that enqueuing has succeeded.
+		go sendResponseWithDelay(f, 2*delayResponse, "test", msg.QueryID, &httpgrpc.HTTPResponse{
+			Code: 200,
+			Body: []byte("ok"),
+		})
+
+		// give enough time to fill up inflight requests
+		time.Sleep(delayResponse)
+		return &schedulerpb.SchedulerToFrontend{Status: schedulerpb.OK}
+	})
+
+	inflightRequests := 10
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for i := 0; i < inflightRequests; i++ {
+		go func() {
+			_, err := f.RoundTripGRPC(user.InjectOrgID(ctx, "test"), &httpgrpc.HTTPRequest{})
+			require.NoError(t, err)
+		}()
+	}
+
+	require.Eventually(t, func() bool {
+		return f.requests.count() == inflightRequests
+	},
+		3*delayResponse, // wait at least 2*delayResponse to wait for queries to be finished and removed from inlight requests map
+		5*time.Millisecond,
+	)
+
+	// blocks until all inflight requests are done
+	_ = f.stopping(nil)
+	require.Equal(t, 0, f.requests.count())
+}
+
+func TestFrontendShuttingDownLetsSubRequestsPass(t *testing.T) {
+	delayResponse := 100 * time.Millisecond
+	f, _ := setupFrontend(t, func(f *Frontend, msg *schedulerpb.FrontendToScheduler) *schedulerpb.SchedulerToFrontend {
+		// We cannot call QueryResult directly, as Frontend is not yet waiting for the response.
+		// It first needs to be told that enqueuing has succeeded.
+		go sendResponseWithDelay(f, delayResponse, "test", msg.QueryID, &httpgrpc.HTTPResponse{
+			Code: 200,
+			Body: []byte("ok"),
+		})
+
+		return &schedulerpb.SchedulerToFrontend{Status: schedulerpb.OK}
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	require.Equal(t, services.Running, f.State())
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		_, err := f.RoundTripGRPC(user.InjectOrgID(ctx, "test"), &httpgrpc.HTTPRequest{})
+		require.NoError(t, err)
+	}()
+
+	// Wait less than delayResponse to make sure we have an inflight request that
+	// already was sent to the scheduler and the service stays in Stopping state
+	// for some time.
+	time.Sleep(delayResponse / 10)
+
+	f.StopAsync()
+
+	// wait for Stopping state
+	require.Eventually(t, func() bool {
+		t.Log(f.State())
+		return f.State() == services.Stopping
+	}, delayResponse/2, 2*time.Millisecond)
+
+	// send (sub-)request
+	// This request still needs to be able to pass the RoundTripGRCP function,
+	// because it may be a sub-request of a query request that was split earlier
+	// into multiple sub-requests.
```

Comment:
why do we need it? I believe that another query-frontend can handle it... 
if we started the shutdown process for this query-frontend, it should not receive any new requests... Wdyt?

Reply:
It's a test for the very counter-intuitive behaviour. By intuition, you would say that the RoundTripGRCP must not access any more incoming requests if the state is not `Running`. However, the opposite is the case, because the function is not the entry-point for requests, but rather a grcp round tripper that is wrapped into the query splitting tripperware, that produces potentially lots of internal gRPC requests that are handled by this function.

Reply:
ok, i see.